torchinductor/lowering.py

```diff
@@ -1305,6 +1305,55 @@ def fn(idx):
     )
 
 
+@register_lowering(aten.scatter_add, type_promote=False)
+def scatter_add(x, dim: int, index, src):
+    assert isinstance(x, TensorBox)
+    assert isinstance(dim, int)
+    assert "int" in str(index.get_dtype())
+    assert 0 <= dim < len(x.get_size())
+
+    index_loader = index.make_loader()
+    src_loader = src.make_loader()
+
+    def output_indexer(idx):
+        indirect_idx = list(idx)
+        indirect_idx[dim] = ops.indirect_indexing(index_loader(idx))
+        return indirect_idx
+
+    def fn(idx):
+        for i in range(len(idx)):
+            boundary_check = ops.lt(
+                ops.index_expr(idx[i], torch.int64),
+                ops.index_expr(index.get_size()[i], torch.int64),
+            )
+            if i == 0:
+                condition = boundary_check
+            else:
+                condition = ops.and_(condition, boundary_check)
+        return ops.where(
+            condition,
+            src_loader(idx),
+            ops.constant(0, x.get_dtype()),
+        )
+
+    result = clone(x)
+    scatter = ir.Scatter(
+        device=x.get_device(),
+        dtype=x.get_dtype(),
+        inner_fn=fn,
+        ranges=x.get_size(),
```

Comment:
Should ranges be `index.get_size()`?

If we are iterating over the ranges of `x` I would expect we need to broadcast `index` into the same shape as `x`.  I haven't looked up the semantics of scatter_add.

Reply:
https://pytorch.org/docs/stable/generated/torch.Tensor.scatter_add_.html#torch.Tensor.scatter_add_ explicitly says `index` and `src` do not broadcast.

Reply:
How many total adds are performed?  If it is `index.numel()` then this line should be `index.get_size()`.

Reply:
Yes, you are right. The cpu backend crash was a SEGFAULT caused by this. Fixed in the new commit.

There is still one problem remaining. When `index` only contains one element, the generated Triton code produces incorrect result. I vaguely remember seeing you have filed some related Triton issue. If not, I can file one.